ledger-service/http-json/src/main/scala/com/digitalasset/http/dbbackend/ContractDao.scala

```diff
@@ -36,40 +37,50 @@ object ContractDao {
   def initialize(implicit log: LogHandler): ConnectionIO[Unit] =
     Queries.dropAllTablesIfExist *> Queries.initDatabase
 
-  def lastOffset(party: domain.Party, templateId: domain.TemplateId.RequiredPkg)(
-      implicit log: LogHandler): ConnectionIO[Option[domain.Offset]] =
+  def lastOffset(parties: OneAnd[Set, domain.Party], templateId: domain.TemplateId.RequiredPkg)(
+      implicit log: LogHandler): ConnectionIO[Map[domain.Party, domain.Offset]] = {
+    import doobie.postgres.implicits._
     for {
       tpId <- Queries.surrogateTemplateId(
         templateId.packageId,
         templateId.moduleName,
         templateId.entityName)
-      offset <- Queries.lastOffset(party.unwrap, tpId).map(_.map(domain.Offset(_)))
+      offset <- Queries
+        .lastOffset(domain.Party.unsubst(parties), tpId)
+        .map(_.map { case (k, v) => (domain.Party(k), domain.Offset(v)) })
     } yield offset
```

Comment:
Given `Map[l, r]` where you want to subst in each type argument, the right one can be fully inferred, but for the left you must specify a type function. The right function, `[a] => Map[l, a]` is inferred.

You can equally subst into the l and r of `ConnectionIO[Map[l, r]]`, but must specify both type functions in that case, because the lambda over `r` cannot be inferred in this case.

```suggestion
      offset <- Queries
        .lastOffset(domain.Party.unsubst(parties), tpId)
    } yield {
      type L[a] = Map[a, domain.Offset]
      domain.Party.subst[L, String](domain.Offset.tag.subst(offset))
    }
```

Reply:
Thanks!